File: kaji/ts/src/cli/connect.ts
import { GoogleOAuthClient, validateOAuthPrincipal } from "@/auth/oauth";
import { MacOSKeychainTokenStorage } from "@/auth/keychain";
import type { RunOptions } from "@/cli/index";
import { TYPESCRIPT_SDK_CLI } from "@/cli/package-identity";
import { closedRecoveryFields, recoveryForReason } from "@/integrations/recovery";
import {
  formatIntegrationError,
  loadManifest,
  loadRegistryIndex,
  type IntegrationAuth,
  type LoadedIntegrationManifest,
} from "@/integrations/registry-loader";

export type OAuthLoadedManifest = Omit<LoadedIntegrationManifest, "auth"> & {
  readonly auth: Extract<IntegrationAuth, { kind: "oauth" }>;
};

export interface AuthArgs {
  readonly name: string;
  readonly principal: string;
  readonly forceLocal: boolean;
}

export const CONNECT_USAGE = "usage: kaji connect <name> --principal <stable-host-principal-id>";

export function parseAuthArgs(
  rest: readonly string[],
  allowForceLocal: boolean,
): AuthArgs | undefined {
  let name: string | undefined;
  let principal: string | undefined;
  let forceLocal = false;
  for (let index = 0; index < rest.length; index++) {
    const argument = rest[index]!;
    if (argument === "--principal") {
      if (principal !== undefined) return undefined;
      const value = rest[index + 1];
      if (value === undefined || value.startsWith("--")) return undefined;
      principal = value;
      index++;
    } else if (argument === "--force-local" && allowForceLocal) {
      if (forceLocal) return undefined;
      forceLocal = true;
    } else if (argument.startsWith("-") || name !== undefined) {
      return undefined;
    } else {
      name = argument;
    }
  }
  if (name === undefined || principal === undefined) return undefined;
  return { name, principal, forceLocal };
}

export function qualifiedAuthCommand(action: "connect" | "disconnect", name: string): string {
  return `${TYPESCRIPT_SDK_CLI} ${action} ${name} ` + "--principal <stable-host-principal-id>";
}

export async function oauthManifest(name: string, opts: RunOptions): Promise<OAuthLoadedManifest> {
  const index = await loadRegistryIndex(opts.registryRoot, { schemaRoot: opts.schemaRoot });
  const manifest = await loadManifest(opts.registryRoot, name, {
    schemaRoot: opts.schemaRoot,
    index,
  });
  if (manifest.auth.kind !== "oauth") {
    throw new Error(`Integration '${name}' does not use OAuth authentication.`);
  }
  if (manifest.auth.provider !== "google") {
    throw new Error(`Integration '${name}' does not use Google OAuth.`);
  }
  return manifest as OAuthLoadedManifest;
}

export function renderClosedRecovery(
  error: unknown,
  command: string,
  output: (message: string) => void,
): boolean {
  const fields = closedRecoveryFields(error);
  if (fields === undefined) return false;
  const recovery = recoveryForReason(fields.reason_code);
  output(`Problem: ${recovery.problem}`);
  output(`Cause: ${recovery.cause}`);
  output(`Fix: ${recovery.fix}`);
  output(`Command: ${command}`);
  return true;
}

export async function withAuthSignal<T>(
  opts: RunOptions,
  operation: (signal: AbortSignal) => Promise<T>,
): Promise<T> {
  if (opts.signal !== undefined) return operation(opts.signal);
  const controller = new AbortController();
  const cancel = () => controller.abort(new DOMException("CLI auth cancelled", "AbortError"));
  process.once("SIGINT", cancel);
  process.once("SIGTERM", cancel);
  try {
    return await operation(controller.signal);
  } finally {
    process.removeListener("SIGINT", cancel);
    process.removeListener("SIGTERM", cancel);
  }
}

function storage(opts: RunOptions, integrationName: string, principal: string) {
  if (opts.keychainStorageFactory !== undefined) {
    return opts.keychainStorageFactory(integrationName);
  }
  const value = new MacOSKeychainTokenStorage(integrationName);
  (value as unknown as { preflight(principalId: string): string }).preflight(principal);
  return value;
}

function client(opts: RunOptions, options: ConstructorParameters<typeof GoogleOAuthClient>[0]) {
  return opts.googleOAuthClientFactory?.(options) ?? new GoogleOAuthClient(options);
}

export async function connectIntegration(rest: string[], opts: RunOptions): Promise<number> {
  const log = opts.log ?? ((message: string) => console.log(message));
  const err = opts.err ?? ((message: string) => console.error(message));
  const args = parseAuthArgs(rest, false);
  if (args === undefined) {
    err(CONNECT_USAGE);
    return 2;
  }
  let principal: string;
  try {
    principal = validateOAuthPrincipal(args.principal);
  } catch {
    err("INTEGRATION_POLICY_REJECTED: The principal identifier is invalid.");
    return 1;
  }
  let manifest: OAuthLoadedManifest;
  try {
    manifest = await oauthManifest(args.name, opts);
  } catch (error) {
    err(
      error instanceof Error && error.message.includes("OAuth")
        ? error.message
        : formatIntegrationError(error),
    );
    return 1;
  }
  const environment = opts.env ?? process.env;
  const clientId = environment[manifest.auth.clientIdEnv];
  const command = qualifiedAuthCommand("connect", manifest.name);
  if (clientId === undefined || clientId.length === 0) {
    err(`INTEGRATION_AUTH_REQUIRED: ${manifest.auth.clientIdEnv} is not set.`);
    err("Create a Google Desktop OAuth client and load the manifest-declared client ID.");
    err(command);
    return 1;
  }
  const clientSecret =
    manifest.auth.clientSecretEnv === undefined
      ? undefined
      : environment[manifest.auth.clientSecretEnv];
  try {
    const oauth = client(opts, {
      clientId,
      ...(clientSecret === undefined ? {} : { clientSecret }),
      scopes: [...manifest.auth.scopes],
      storage: storage(opts, manifest.name, principal),
    });
    await withAuthSignal(opts, (signal) => oauth.connect(principal, signal));
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      err("Problem: Gmail authorization was cancelled.");
      err("Cause: The connect operation was cancelled before completion.");
      err(`Fix: Rerun \`${command}\`.`);
    } else if (!renderClosedRecovery(error, command, err)) {
      err("Problem: Google OAuth consent did not complete.");
      err("Cause: The provider denied or failed installed-app consent.");
      err(`Fix: Rerun \`${command}\`.`);
    }
    return 1;
  }
  log(`Connected ${manifest.name} for the requested principal.`);
  log(`Stored refresh credentials in macOS Keychain service dev.kaji.oauth.${manifest.name}.`);
  return 0;
}

File: kaji/sdk/src/cli/connect.py
"""`kaji connect`: explicitly establish a manifest-declared OAuth grant."""

from __future__ import annotations

import argparse
import asyncio
from collections.abc import Mapping, Sequence
import os
from typing import Any

from kaji.integrations import (
    IntegrationNotFound,
    Manifest,
    ManifestError,
    load_manifest,
)
from kaji.integrations.errors import IntegrationExecutionError
from kaji.integrations.keychain import MacOSKeychainTokenStorage
from kaji.integrations.oauth import GoogleOAuthClient, OAuthError, _require_principal
from kaji.integrations.recovery import recovery_for_reason
from kaji.runtime.agents.cancellation import CancellationToken


_environment: Mapping[str, str] = os.environ


class UniqueValue(argparse.Action):
    def __call__(
        self,
        parser: argparse.ArgumentParser,
        namespace: argparse.Namespace,
        values: str | Sequence[Any] | None,
        option_string: str | None = None,
    ) -> None:
        if getattr(namespace, self.dest, None) is not None:
            parser.error(f"{option_string} may be specified only once")
        setattr(namespace, self.dest, values)


def add_parser(sub: argparse._SubParsersAction) -> None:
    parser = sub.add_parser("connect", help="connect an integration OAuth grant")
    parser.add_argument("name")
    parser.add_argument("--principal", required=True, action=UniqueValue, default=None)
    parser.set_defaults(func=run)


def qualified_command(name: str, action: str) -> str:
    return f"python -m kaji.cli {action} {name} --principal <stable-host-principal-id>"


def oauth_manifest(name: str) -> Manifest:
    manifest = load_manifest(name)
    if manifest.auth.kind != "oauth":
        raise ValueError(f"Integration {name!r} does not use OAuth.")
    if manifest.auth.provider != "google":
        raise ValueError(f"Integration {name!r} has an unsupported OAuth provider.")
    return manifest


def _production_client(
    *,
    manifest: Manifest,
    principal: str,
    client_id: str | None,
    client_secret: str | None,
) -> GoogleOAuthClient:
    storage = MacOSKeychainTokenStorage(manifest.name)
    storage._preflight(principal)
    return GoogleOAuthClient(
        client_id=client_id,
        client_secret=client_secret,
        scopes=manifest.auth.scopes,
        credential_store=storage,
    )


def render_recovery(error: IntegrationExecutionError, *, command: str) -> None:
    recovery = recovery_for_reason(error.reason_code)
    print(recovery.error_code)
    print(f"Problem: {recovery.problem}")
    print(f"Cause: {recovery.cause}")
    print(f"Fix: {recovery.fix}")
    print(f"Command: {command}")


async def _connect(client: GoogleOAuthClient, principal: str) -> None:
    await client.connect(principal, CancellationToken())


def run(args: argparse.Namespace) -> int:
    command = qualified_command(args.name, "connect")
    try:
        principal = _require_principal(args.principal)
    except IntegrationExecutionError:
        print("INTEGRATION_POLICY_REJECTED: The principal identifier is invalid.")
        print(f"Command: {command}")
        return 1
    try:
        manifest = oauth_manifest(args.name)
    except IntegrationNotFound:
        print(f"Unknown integration: {args.name!r}.")
        return 1
    except (ManifestError, ValueError) as error:
        print(str(error))
        return 1

    client_id_name = manifest.auth.client_id_env
    if client_id_name is None:
        print("INTEGRATION_AUTH_REQUIRED: OAuth client ID metadata is missing.")
        print(f"Command: {command}")
        return 1
    client_id = _environment.get(client_id_name)
    if not client_id:
        print(f"INTEGRATION_AUTH_REQUIRED: {client_id_name} is not set.")
        print(
            f"Create a Google Desktop OAuth client, load {client_id_name}, then rerun:"
        )
        print(command)
        return 1
    client_secret = (
        _environment.get(manifest.auth.client_secret_env)
        if manifest.auth.client_secret_env is not None
        else None
    )
    try:
        client = _production_client(
            manifest=manifest,
            principal=principal,
            client_id=client_id,
            client_secret=client_secret,
        )
        asyncio.run(_connect(client, principal))
    except asyncio.CancelledError:
        print("INTEGRATION_AUTH_ERROR: OAuth connection was cancelled.")
        print("Cause: The operation was cancelled before grant storage completed.")
        print(f"Command: {command}")
        return 1
    except KeyboardInterrupt:
        print("INTEGRATION_AUTH_ERROR: OAuth connection was cancelled.")
        print("Cause: The operation was interrupted before grant storage completed.")
        print(f"Command: {command}")
        return 1
    except IntegrationExecutionError as error:
        render_recovery(error, command=command)
        return 1
    except OAuthError:
        print("INTEGRATION_AUTH_ERROR: Google OAuth consent did not complete.")
        print("Cause: The provider denied or failed the installed-app consent flow.")
        print(f"Command: {command}")
        return 1

    print(f"Connected {manifest.name} for the requested principal.")
    print(
        "Stored refresh credentials in macOS Keychain service "
        f"dev.kaji.oauth.{manifest.name}."
    )
    return 0
